Add terminal npm script for env variables

adelineu opened the pull request:
Rewrites default arguments or adds new ones.
Given arguments are being saved in .with.json in root directory.

Example:
npm run with server:domain=google.com server:port=80 is-dev browser=false && npm run commands

Insight:
- server:domain=google.com gets translated to SERVER_DOMAIN and the value "google.com" - available via process.env.SERVER_DOMAIN
- is-dev to DEV- available via process.env.DEV
- browser to BROWSER - available via process.env.BROWSER

morion4000 commented on build_webpack/webpack.terminal.config.js:
```diff
@@ -2,11 +2,26 @@ const webpack = require('webpack');
 const nodeExternals = require('webpack-node-externals');
+const addWithsConfig = require('../scripts/npm/with');
+
+const default_arguments = {
```
How are the default_arguments handled? Are they overwritten by the command line arguments?

adelineu replied:
Yes.

If you got a key in default_arguments, it gets overwritten by the arguments passed in with's command line.

morion4000 commented on build_webpack/webpack.terminal.config.js:
```diff
@@ -2,11 +2,26 @@ const webpack = require('webpack');
+const default_arguments = {
+    BROWSER: 'false',
+    SERVER_PORT: process.env.SERVER_PORT,
+    PORT: process.env.PORT,
```
here you have `PORT` and in the command line argument you have `SERVER_PORT`. is there a conflict?

adelineu replied:
I don't know. That's how it looked by default.

morion4000 replied:
Ok, I will look into it.

morion4000 commented on scripts/npm/with/storage.js:
```diff
@@ -0,0 +1,27 @@
+const FS = require('fs');
+
+class Storage {
```
I do not understand the purpose of the storage class. Is it so that you save the parameters in a file so you don't have to write then every time?

adelineu replied:
I didn't find it makes sense that saving the given parameters is parser's (arguments parser's) job.
So, Storage class is just a small abstraction.

morion4000 commented on .gitignore:
```diff
@@ -68,4 +68,6 @@ transactionsDB/
\ No newline at end of file
+certificates/private.key
+
+.with.json
```
What does this file do?

adelineu replied:
holds the give parameters as json.

A package script like "npm run terminal" has other commands in it and we can't pass arguments dynamically to those commands.

So we needed a way to load up those given arguments in other given commands.

npm run with server:domain=google.com has as purpose just to parse those arguments and save them to that file.
npm run commands checks to see if that file exists. If it exists and it has data in it, then it gets loaded. If there are arguments that are in given defaults (as i stated above) then they get overwritten by the ones from the file.

morion4000 commented:
Does `is-dev` sets `process.env.DEV` to `true`?

morion4000 commented:
How easy is to create a new parameter? Something like: `TERMINAL_WORKERS`.`MAX` for example.

adelineu commented:
New version is now available:

- cleared the code that was used as example from build_webpack/webpack.terminal.config.js
so now it's on his original version and it needs to be edited in order to see this functional.
```
const savedConfig = require('../scripts/npm/with');
const default_arguments = {};
const envData = savedConfig(default_arguments);

const config = merge(base, {
    plugins: [
        new webpack.DefinePlugin({
            "env": envData
        })
    ]
});
```
- fixed the digit regex
- moved the default expressions to Parser
- improved namespacing and now a command like npm run with server:port=80 server:domain="google.com" ends interpreted internally like
SERVER: {PORT: 80, domain: "google.com"} and values are available via something like env.SERVER.PORT (_see the note_)

NOTE:
For some reason if you use process.env as a key for WebpackPlugin it's acts a bit fishy. If you want to use this in production i suggest renaming it to something like env

adelineu commented:
Code was rebased.
Made public both read and save methods.

```
    // simple access to read & save
    module.exports = {
        read: (data, source) => {
            return (new Storage(source)).read(data);
        },
        save: (data, source) => {
            return (new Storage(source)).save(data);
        }
    };
```

Usage has been updated in previous comment to:
`const envData = savedConfig.read(default_arguments);`

adelineu commented:
wrote new version. check PR 90
